File: jvm/selfie-lib/src/main/kotlin/com/diffplug/selfie/ArrayMap.kt
/*
 * Copyright (C) 2023-2024 DiffPlug
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.diffplug.selfie

import kotlin.collections.binarySearch
private val STRING_SLASHFIRST =
    Comparator<String> { a, b ->
      var i = 0
      while (i < a.length && i < b.length) {
        val charA = a[i]
        val charB = b[i]
        if (charA != charB) {
          return@Comparator ( // treat
          if (charA == '/') -1 // slash as
          else if (charB == '/') 1 // the lowest
          else charA.compareTo(charB) // character
          )
        }
        i++
      }
      a.length.compareTo(b.length)
    }
private val PAIR_STRING_SLASHFIRST =
    Comparator<Pair<String, Any>> { a, b -> STRING_SLASHFIRST.compare(a.first, b.first) }

abstract class ListBackedSet<T>() : AbstractSet<T>() {
  abstract operator fun get(index: Int): T
  override fun iterator(): Iterator<T> =
      object : Iterator<T> {
        private var index = 0
        override fun hasNext(): Boolean = index < size
        override fun next(): T {
          if (!hasNext()) throw NoSuchElementException()
          return get(index++)
        }
      }
}
private fun <T : Comparable<T>> ListBackedSet<T>.binarySearch(element: T): Int {
  val list =
      object : AbstractList<T>() {
        override val size: Int
          get() = this@binarySearch.size
        override fun get(index: Int): T = this@binarySearch[index]
      }
  return if (element is String) {
    (list as List<String>).binarySearch(element, STRING_SLASHFIRST)
  } else list.binarySearch(element)
}

internal expect fun <K, V> entry(key: K, value: V): Map.Entry<K, V>

/**
 * An immutable, sorted, array-backed map - if keys are [String], they have a special sort order
 * where `/` is the lowest key.
 */
class ArrayMap<K : Comparable<K>, V : Any>(private val data: Array<Any>) : Map<K, V> {
  fun minusSortedIndices(indicesToRemove: List<Int>): ArrayMap<K, V> {
    if (indicesToRemove.isEmpty()) {
      return this
    }
    val newData = arrayOfNulls<Any>(data.size - indicesToRemove.size * 2)
    var newDataIdx = 0
    var currentPairIdx = 0 // Index for key-value pairs
    var toRemoveIdx = 0
    while (currentPairIdx < data.size / 2) {
      if (toRemoveIdx < indicesToRemove.size && currentPairIdx == indicesToRemove[toRemoveIdx]) {
        toRemoveIdx++
      } else {
        check(newDataIdx < newData.size) {
          "The indices weren't sorted or were >= size=$size: $indicesToRemove"
        }
        newData[newDataIdx++] = data[2 * currentPairIdx] // Copy key
        newData[newDataIdx++] = data[2 * currentPairIdx + 1] // Copy value
      }
      currentPairIdx++
    }

    // Ensure all indices to remove have been processed
    check(toRemoveIdx == indicesToRemove.size) {
      "The indices weren't sorted or were >= size($size): $indicesToRemove"
    }
    return ArrayMap<K, V>(newData as Array<Any>)
  }

  /**
   * Returns a new ArrayMap which has added the given key. Throws an exception if the key already
   * exists.
   */
  fun plus(key: K, value: V): ArrayMap<K, V> {
    val next = plusOrNoOp(key, value)
    if (next === this) {
      throw IllegalArgumentException("Key already exists: $key")
    }
    return next
  }
  /**
   * Returns a new ArrayMap which has added the given key, or the current map if the key was already
   * in the map.
   */
  fun plusOrNoOp(key: K, value: V): ArrayMap<K, V> {
    val idxExisting = dataAsKeys.binarySearch(key)
    return if (idxExisting >= 0) this
    else {
      val idxInsert = -(idxExisting + 1)
      insert(idxInsert, key, value)
    }
  }
  /**
   * Returns an ArrayMap which has added or overwritten the given key/value. If the map already
   * contained that mapping (equal keys and values) then it returns the identically same map.
   */
  fun plusOrNoOpOrReplace(key: K, newValue: V): ArrayMap<K, V> {
    val idxExisting = dataAsKeys.binarySearch(key)
    if (idxExisting >= 0) {
      val existingValue = data[idxExisting * 2 + 1] as V
      return if (newValue == existingValue) this
      else {
        val copy = data.copyOf()
        copy[idxExisting * 2 + 1] = newValue
        return ArrayMap(copy)
      }
    } else {
      val idxInsert = -(idxExisting + 1)
      return insert(idxInsert, key, newValue)
    }
  }
  private fun insert(idxInsert: Int, key: K, value: V): ArrayMap<K, V> {
    return when (data.size) {
      0 -> ArrayMap(arrayOf(key, value))
      1 -> {
        if (idxInsert == 0) ArrayMap(arrayOf(key, value, data[0], data[1]))
        else ArrayMap(arrayOf(data[0], data[1], key, value))
      }
      else ->
          // TODO: use idxInsert and arrayCopy to do this faster, see ArraySet#plusOrThis
          of(
              MutableList(size + 1) {
                if (it < size) Pair(data[it * 2] as K, data[it * 2 + 1] as V) else Pair(key, value)
              })
    }
  }
  /** list-backed set of guaranteed-sorted keys at even indices. */
  private val dataAsKeys =
      object : ListBackedSet<K>() {
        override val size: Int
          get() = data.size / 2
        override fun get(index: Int): K {
          return data[index * 2] as K
        }
      }
  override val keys: ListBackedSet<K>
    get() = dataAsKeys
  override fun get(key: K): V? {
    val idx = dataAsKeys.binarySearch(key)
    return if (idx < 0) null else data[idx * 2 + 1] as V
  }
  override fun containsKey(key: K): Boolean = dataAsKeys.binarySearch(key) >= 0
  /** list-backed collection of values at odd indices. */
  override val values: List<V>
    get() =
        object : AbstractList<V>() {
          override val size: Int
            get() = data.size / 2
          override fun get(index: Int): V = data[index * 2 + 1] as V
        }
  override fun containsValue(value: V): Boolean = values.contains(value)
  /** list-backed set of entries. */
  override val entries: ListBackedSet<Map.Entry<K, V>>
    get() =
        object : ListBackedSet<Map.Entry<K, V>>() {
          override val size: Int
            get() = data.size / 2
          override fun get(index: Int): Map.Entry<K, V> {
            val key = data[index * 2] as K
            val value = data[index * 2 + 1] as V
            return entry(key, value)
          }
        }
  override val size: Int
    get() = data.size / 2
  override fun isEmpty(): Boolean = data.isEmpty()
  override fun equals(other: Any?): Boolean {
    if (other === this) return true
    if (other !is Map<*, *>) return false
    if (size != other.size) return false
    return if (other is ArrayMap<*, *>) {
      data.contentEquals(other.data)
    } else {
      other.entries.all { (key, value) -> this[key] == value }
    }
  }
  override fun hashCode(): Int = entries.hashCode()
  override fun toString() = this.toMutableMap().toString()

  companion object {
    private val EMPTY = ArrayMap<String, Any>(arrayOf())
    fun <K : Comparable<K>, V : Any> empty() = EMPTY as ArrayMap<K, V>
    fun <K : Comparable<K>, V : Any> of(pairs: MutableList<Pair<K, V>>): ArrayMap<K, V> {
      if (pairs.size > 1) {
        if (pairs[0].first is String) {
          (pairs as (MutableList<Pair<String, Any>>)).sortWith(PAIR_STRING_SLASHFIRST)
        } else {
          pairs.sortBy { it.first }
        }
      }
      val array = arrayOfNulls<Any>(pairs.size * 2)
      for (i in 0 until pairs.size) {
        array[i * 2] = pairs[i].first
        array[i * 2 + 1] = pairs[i].second
      }
      return ArrayMap(array as Array<Any>)
    }
  }
}

/**
 * An immutable, sorted, array-backed set - if keys are [String], they have a special sort order
 * where `/` is the lowest key.
 */
class ArraySet<K : Comparable<K>>(private val data: Array<Any>) : ListBackedSet<K>() {
  override val size: Int
    get() = data.size
  override fun get(index: Int): K = data[index] as K
  override fun contains(element: K): Boolean = binarySearch(element) >= 0
  fun plusOrThis(key: K): ArraySet<K> {
    val idxExisting = binarySearch(key)
    if (idxExisting >= 0) {
      return this
    }
    val idxInsert = -(idxExisting + 1)
    return when (data.size) {
      0 -> ArraySet(arrayOf(key))
      1 -> {
        if (idxInsert == 0)
            ArraySet(
                arrayOf(
                    key,
                    data[0],
                ))
        else ArraySet(arrayOf(data[0], key))
      }
      else -> {
        // TODO: use idxInsert and arrayCopy to do this faster, see ArrayMap#insert
        val array = Array(size + 1) { if (it < size) data[it] else key }
        if (key is String) {
          array.sortWith(STRING_SLASHFIRST as Comparator<Any>)
        } else {
          (array as Array<K>).sort()
        }
        ArraySet(array)
      }
    }
  }

  companion object {
    private val EMPTY = ArraySet<String>(arrayOf())
    fun <K : Comparable<K>> empty() = EMPTY as ArraySet<K>
  }
}

File: dotnet/Selfie.Lib/ArrayMap.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;

namespace DiffPlug.Selfie.Lib;

public abstract class ListBackedSet<T> : ISet<T> {
  public abstract T this[int index] { get; }
  public abstract int Count { get; }
  public bool IsReadOnly => false;

  public IEnumerator<T> GetEnumerator() {
    for (var i = 0; i < Count; i++) {
      yield return this[i];
    }
  }

  IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

  public bool Contains(T item) => IndexOf(item) >= 0;

  public void CopyTo(T[] array, int arrayIndex) {
    for (var i = 0; i < Count; i++) {
      array[arrayIndex + i] = this[i];
    }
  }

  public bool Add(T item) => throw new NotSupportedException();
  public void Clear() => throw new NotSupportedException();
  public bool Remove(T item) => throw new NotSupportedException();

  public void ExceptWith(IEnumerable<T> other) => throw new NotSupportedException();
  public void IntersectWith(IEnumerable<T> other) => throw new NotSupportedException();
  public void SymmetricExceptWith(IEnumerable<T> other) => throw new NotSupportedException();
  public void UnionWith(IEnumerable<T> other) => throw new NotSupportedException();

  public int IndexOf(T item) {
    var comparer = GetComparer(item);
    for (var i = 0; i < Count; i++) {
      if (comparer.Compare(this[i], item) == 0) {
        return i;
      }
    }
    return -1;
  }

  public bool IsProperSubsetOf(IEnumerable<T> other) {
    var otherSet = new HashSet<T>(other);
    return Count < otherSet.Count && IsSubsetOf(otherSet);
  }

  public bool IsProperSupersetOf(IEnumerable<T> other) {
    var otherSet = new HashSet<T>(other);
    return Count > otherSet.Count && otherSet.IsSubsetOf(this);
  }

  public bool IsSubsetOf(IEnumerable<T> other) {
    var otherSet = new HashSet<T>(other);
    return this.All(item => otherSet.Contains(item));
  }

  public bool IsSupersetOf(IEnumerable<T> other) {
    var otherSet = new HashSet<T>(other);
    return otherSet.IsSubsetOf(this);
  }

  public bool Overlaps(IEnumerable<T> other) {
    return other.Any(Contains);
  }

  public bool SetEquals(IEnumerable<T> other) {
    var otherSet = new HashSet<T>(other);
    return Count == otherSet.Count && IsSubsetOf(otherSet);
  }

  void ICollection<T>.Add(T item) => Add(item);

  private static IComparer<T> GetComparer(T element) =>
      typeof(T) == typeof(string) ? (IComparer<T>)new StringComparer() : Comparer<T>.Default;

  private class StringComparer : IComparer<string> {
    public int Compare(string? x, string? y) => CompareStringsWithSlashFirst(x, y);
  }

  protected static int CompareStringsWithSlashFirst(string? a, string? b) {
    if (a == b) {
      return 0;
    }

    if (a == null) {
      return -1;
    }

    if (b == null) {
      return 1;
    }

    var length = Math.Min(a.Length, b.Length);
    for (var i = 0; i < length; i++) {
      var charA = a[i];
      var charB = b[i];
      if (charA != charB) {
        return charA == '/' ? -1 :
               charB == '/' ? 1 :
               charA.CompareTo(charB);
      }
    }
    return a.Length.CompareTo(b.Length);
  }
}

public class ArrayMap<TKey, TValue> : IDictionary<TKey, TValue>
    where TKey : notnull, IComparable<TKey> {
  private readonly object[] _data;

  private ArrayMap(object[] data) {
    _data = data;
  }

  public ArrayMap<TKey, TValue> MinusSortedIndices(IReadOnlyList<int> indicesToRemove) {
    if (indicesToRemove.Count == 0) {
      return this;
    }

    var newData = new object[_data.Length - indicesToRemove.Count * 2];
    var newDataIdx = 0;
    var oldDataIdx = 0;
    var removeIdx = 0;

    while (oldDataIdx < _data.Length / 2) {
      if (removeIdx < indicesToRemove.Count && oldDataIdx == indicesToRemove[removeIdx]) {
        removeIdx++;
      }
      else {
        if (newDataIdx >= newData.Length) {
          throw new ArgumentException(
              $"The indices weren't sorted or were >= Count ({Count}): {string.Join(", ", indicesToRemove)}");
        }
        newData[newDataIdx++] = _data[oldDataIdx * 2];
        newData[newDataIdx++] = _data[oldDataIdx * 2 + 1];
      }
      oldDataIdx++;
    }

    if (removeIdx != indicesToRemove.Count) {
      throw new ArgumentException(
          $"The indices weren't sorted or were >= Count ({Count}): {string.Join(", ", indicesToRemove)}");
    }

    return new ArrayMap<TKey, TValue>(newData);
  }

  public ArrayMap<TKey, TValue> Plus(TKey key, TValue value) {
    var next = PlusOrNoOp(key, value);
    if (next == this) {
      throw new ArgumentException($"Key already exists: {key}");
    }
    return next;
  }

  public ArrayMap<TKey, TValue> PlusOrNoOp(TKey key, TValue value) {
    var index = KeysList.IndexOf(key);
    return index >= 0 ? this : Insert(~index, key, value);
  }

  public ArrayMap<TKey, TValue> PlusOrNoOpOrReplace(TKey key, TValue newValue) {
    var index = KeysList.IndexOf(key);
    if (index >= 0) {
      var existingValue = _data[index * 2 + 1];
      return existingValue == null && newValue == null || existingValue?.Equals(newValue) == true
          ? this
          : ReplaceValue(index, newValue);
    }
    else {
      return Insert(~index, key, newValue);
    }
  }

  private ArrayMap<TKey, TValue> Insert(int index, TKey key, TValue value) {
    switch (_data.Length) {
      case 0:
        return new ArrayMap<TKey, TValue>(new object[] { key!, value! });
      case 1:
        return index == 0
            ? new ArrayMap<TKey, TValue>(new object[] { key!, value!, _data[0]!, _data[1]! })
            : new ArrayMap<TKey, TValue>(new object[] { _data[0]!, _data[1]!, key!, value! });
      default: {
          var pairs = new KeyValuePair<TKey, TValue>[Count + 1];
          for (var i = 0; i < index; i++) {
            pairs[i] = new KeyValuePair<TKey, TValue>((TKey)_data[i * 2]!, (TValue)_data[i * 2 + 1]!);
          }
          pairs[index] = new KeyValuePair<TKey, TValue>(key, value);
          for (var i = index + 1; i < pairs.Length; i++) {
            pairs[i] = new KeyValuePair<TKey, TValue>((TKey)_data[(i - 1) * 2]!, (TValue)_data[(i - 1) * 2 + 1]!);
          }
          return Of(pairs.ToArray());
        }
    }
  }

  private ArrayMap<TKey, TValue> ReplaceValue(int index, TValue newValue) {
    var copy = new object[_data.Length];
    Array.Copy(_data, copy, _data.Length);
    copy[index * 2 + 1] = newValue!;
    return new ArrayMap<TKey, TValue>(copy);
  }


  public ListBackedSet<TKey> KeysList => new ArrayMapKeySet(_data);
  public ICollection<TKey> Keys => KeysList;
  public ICollection<TValue> Values => new ArrayMapValueCollection(_data);

  public int Count => _data.Length / 2;
  public bool IsReadOnly => true;

  public TValue this[TKey key] {
    get {
      var index = KeysList.IndexOf(key);
      return index >= 0 ? (TValue)_data[index * 2 + 1]! : throw new KeyNotFoundException(key.ToString());
    }
    set => throw new NotSupportedException();
  }

  public bool ContainsKey(TKey key) => KeysList.IndexOf(key) >= 0;
  public bool TryGetValue(TKey key, out TValue value) {
    var index = KeysList.IndexOf(key);
    if (index >= 0) {
      value = (TValue)_data[index * 2 + 1]!;
      return true;
    }
    else {
      value = default!;
      return false;
    }
  }

  public void Add(TKey key, TValue value) => throw new NotSupportedException();
  public bool Remove(TKey key) => throw new NotSupportedException();
  public void Clear() => throw new NotSupportedException();
  public void Add(KeyValuePair<TKey, TValue> item) => throw new NotSupportedException();

  public bool Remove(KeyValuePair<TKey, TValue> item) => throw new NotSupportedException();

  public IEnumerator<KeyValuePair<TKey, TValue>> GetEnumerator() {
    for (var i = 0; i < Count; i++) {
      yield return new KeyValuePair<TKey, TValue>((TKey)_data[i * 2]!, (TValue)_data[i * 2 + 1]!);
    }
  }

  IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

  public void CopyTo(KeyValuePair<TKey, TValue>[] array, int arrayIndex) {
    for (var i = 0; i < Count; i++) {
      array[arrayIndex + i] = new KeyValuePair<TKey, TValue>((TKey)_data[i * 2]!, (TValue)_data[i * 2 + 1]!);
    }
  }

  public bool Contains(KeyValuePair<TKey, TValue> item) =>
      TryGetValue(item.Key, out var value) &&
      (value == null && item.Value == null || value?.Equals(item.Value) == true);

  public override bool Equals(object? obj) =>
      obj is ArrayMap<TKey, TValue> other && Equals(other);

  public bool Equals(ArrayMap<TKey, TValue>? other) {
    if (other == null || Count != other.Count) {
      return false;
    }

    for (var i = 0; i < Count; i++) {
      if (!Equals(_data[i * 2], other._data[i * 2]) ||
          !Equals(_data[i * 2 + 1], other._data[i * 2 + 1])) {
        return false;
      }
    }
    return true;
  }

  public override int GetHashCode() {
    var hash = 0;
    for (var i = 0; i < _data.Length; i++) {
      hash ^= _data[i]?.GetHashCode() ?? 0;
    }
    return hash;
  }

  public override string ToString() =>
      $"[{string.Join(", ", this.Select(kv => $"{kv.Key}={kv.Value}"))}]";

  private class ArrayMapKeySet : ListBackedSet<TKey> {
    private readonly object[] _data;

    public ArrayMapKeySet(object[] data) {
      _data = data;
    }

    public override TKey this[int index] => (TKey)_data[index * 2]!;
    public override int Count => _data.Length / 2;
  }

  private class ArrayMapValueCollection : ICollection<TValue> {
    private readonly object[] _data;

    public ArrayMapValueCollection(object[] data) {
      _data = data;
    }

    public int Count => _data.Length / 2;
    public bool IsReadOnly => true;

    public IEnumerator<TValue> GetEnumerator() {
      for (var i = 1; i < _data.Length; i += 2) {
        yield return (TValue)_data[i]!;
      }
    }

    IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();

    public bool Contains(TValue item) {
      for (var i = 1; i < _data.Length; i += 2) {
        var value = (TValue)_data[i]!;
        if (value == null && item == null || value?.Equals(item) == true) {
          return true;
        }
      }
      return false;
    }

    public void CopyTo(TValue[] array, int arrayIndex) {
      for (var i = 1; i < _data.Length; i += 2) {
        array[arrayIndex++] = (TValue)_data[i]!;
      }
    }

    public void Add(TValue item) => throw new NotSupportedException();
    public bool Remove(TValue item) => throw new NotSupportedException();
    public void Clear() => throw new NotSupportedException();
  }

  private static readonly ArrayMap<TKey, TValue> EmptyImpl =
      new(Array.Empty<object>());

  public static ArrayMap<TKey, TValue> Empty { get; } = EmptyImpl;

  public static ArrayMap<TKey, TValue> Of(params KeyValuePair<TKey, TValue>[] pairs) {
    if (pairs.Length <= 1) {
      return pairs.Length == 0 ? Empty : new ArrayMap<TKey, TValue>(new object[] { pairs[0].Key!, pairs[0].Value! });
    }

    Array.Sort(pairs, PairComparer.Instance);

    var data = new object[pairs.Length * 2];
    for (var i = 0; i < pairs.Length; i++) {
      data[i * 2] = pairs[i].Key!;
      data[i * 2 + 1] = pairs[i].Value!;
    }
    return new ArrayMap<TKey, TValue>(data);
  }

  private class PairComparer : IComparer<KeyValuePair<TKey, TValue>> {
    public static readonly PairComparer Instance = new();
    public int Compare(KeyValuePair<TKey, TValue> x, KeyValuePair<TKey, TValue> y) => x.Key.CompareTo(y.Key);
  }
}
